### PR TITLE
Multi threading

### PR DESCRIPTION
**Pull request by Honeybunch**
Changes to Core to support features found in the Game and Graphics libraries.
- Added ThreadsafeQueue
- Small changes to Resourcemanager
- Linux  build warning fixes

### CODE REVIEW

File: include/ht_refcounted.h [hunk 1]
```diff
@@ -93,9 +94,11 @@ namespace Hatchit
             Handle(VarType* varPtr, uint32_t* refCounter, const Guid* name);
 
             //Private members
-            VarType* m_ptr;
-            uint32_t* m_refCount;
+            VarType*    m_ptr;
+            uint32_t*   m_refCount;
             const Guid* m_ID;
+
+            mutable std::mutex m_mutex;
```
Comment (ghost) [inline include/ht_refcounted.h:101]: Shouldn't the actual mutex be with the refcounted object?  If two separate handles access the same object and have different mutexes, then we still have a race condition.

File: source/ht_file.cpp [hunk 2]
```diff
@@ -28,7 +28,7 @@ namespace Hatchit
     {
         namespace
         {
-            std::string GetExtension(const std::string& path, bool wd = true)
+            /*std::string GetExtension(const std::string& path, bool wd = true)
```
Comment (ghost) [inline source/ht_file.cpp]: We could just remove this if we aren't using it anymore

File: source/ht_inisettings.cpp [hunk 3]
```diff
@@ -73,7 +73,7 @@ namespace Hatchit
                 return;
 
             auto it = m_values.rbegin();
```
Comment (ghost) [inline source/ht_inisettings.cpp:75]: Why have the declaration outside the for loop?  Looks unorganized.

File: source/inline/ht_jsonhelper.inl [hunk 4]
```diff
@@ -88,6 +88,7 @@ namespace Hatchit
             //Removed for GCC compilation. 
             //TODO: Find cross platform solution.
             //static_assert(false, "Extraction is not defined for this type!");
+            return false;
```
Comment (ghost) [inline source/inline/ht_jsonhelper.inl:91]: Might be able to do `static_assert(sizeof(T) == 0 && false, "Extraction is not defined for this type!");`

File: include/ht_refcounted.h (hunk 1 above)
Comment (MattGuerrette) [inline include/ht_refcounted.h:101]: Yes, the handle should not have the mutex. the refcount could be an atomic, and the guid could be readonly anyway. Its really the refcounted object that needs to be threadsafe.

File: source/ht_file.cpp (hunk 2 above)
Comment (Honeybunch) [inline source/ht_file.cpp]: The commented one? Yeah probably

File: source/inline/ht_jsonhelper.inl (hunk 4 above)
Comment (Honeybunch) [inline source/inline/ht_jsonhelper.inl:91]: @MattGuerrette can you try that and confirm it works?

**Pull request merged**